Add `#[pyo3(deprecated(...))]` to `#[pyfunction]` (#4316)

FlickerSoul opened the pull request:
This PR implements `#[pyo3(deprecated(message = "...", category = ...))]` attribute for `#[pyfunction]` and `#[pymethods]`. The new attribute works like the following 

```rust
#[pyfunction]
#[pyo3(deprecated(message = "this is a deprecated function"))]
fn deprecated_function() {
    // the execution of this function will allow user to see 
    // DeprecationWarning: this is a deprecated function
}

#[pyfunction]
#[pyo3(deprecated(message = "this is a deprecated function", category = PyFutureWarning))]
fn deprecated_function() {
    // and with a specified category
    // PyFutureWarning: This is a warning message
}
```

Note that this function does not deprecate rust function. Programmers need to add `#[deprecated]` attribute explicitly. 

Ref: (#4316)

TODO: 
 - docs to all new functions and / or detail in the guide
 - tests for all new or changed functions

Icxolu commented on pyo3-macros-backend/src/method.rs:
```diff
@@ -740,6 +744,23 @@ impl<'a> FnSpec<'a> {
+            let category = &d.category;
+            match category {
+                Some(category) => {
+                    quote! {
+                        py.import_bound("warnings")?.call_method1("warn", (#message, <#category as #pyo3_path::PyTypeInfo>::type_object_bound(py)))?;
+                    }
```
I think we should emit calls to `PyErr::warn_bound` here, instead of manually executing Python code.

Icxolu commented on pyo3-macros-backend/src/pyfunction.rs:
```diff
@@ -83,13 +83,66 @@ impl PyFunctionArgPyO3Attributes {
+        }
+
+        Ok(PyFunctionDeprecatedAttribute {
+            message: message.expect("`message` is required for `deprecated` attribute"),
```
This should return a `Result::Err` instead of panicking.

Icxolu reviewed (commented):
Thank you very much! I think this looks really good already. Just left some early thoughts while skimming through the macro code.

Icxolu commented on guide/src/function.md:
````diff
@@ -96,6 +96,55 @@ The `#[pyo3]` attribute can be used to modify properties of the generated Python
     ```
+  - <a id="deprecated"></a> `#[pyo3(deprecated(message = "...", category = ...))]`
+
+    Set this option to mark the function as deprecated. The `message` parameter is a string that will be displayed when the function is called, and the `category` parameter is a string that will be displayed in the deprecation warning. The `category` parameter is optional and has to be a subclass of [`Warning`](https://docs.python.org/3.12/library/exceptions.html#Warning). When the `category` par ...
````
`` `category` parameter is a string that will be displayed in the deprecation warning ``

I assume this is some kind of mixup with `message`, at least this does not make a lot of sense to me like this.

Icxolu reviewed (commented):
Implementation wise this looks good to me. I think for testing we still need
- `#[pymethods]` tests, probably in `test_methods.rs`
- coverage in the hygiene tests
- I think this also deserves a `pytest`

For the docs, we might want to say somewhere that this is also available to `#[pymethods]`, but I haven't seen a good place for that yet myself.

FlickerSoul commented:
Thanks for the detailed feedback! I'll work on these points!! :)

mejrs commented on pyo3-macros-backend/src/pyfunction.rs:
```diff
@@ -83,13 +84,105 @@ impl PyFunctionArgPyO3Attributes {
     }
 }
 
+type DeprecatedFunctionMessageAttribute = KeywordAttribute<attributes::kw::message, LitStr>;
```
```suggestion
type DeprecatedFunctionMessageAttribute = KeywordAttribute<attributes::kw::message, LitCStr>;
```
Since we know the string literal at compile time, and `PyErr_WarnEx` takes a null terminated string, we can avoid making the CString in `warn_bound` . We don't have an api for raising warnings that way, so you need to make a function for that as well.

FlickerSoul replied:
I used `PyErr::warn_bound` and that seems to do the conversion to CString and matches what you refer to as as the new API :)

FlickerSoul replied:
The reference is here https://pyo3.rs/main/doc/src/pyo3/err/mod.rs#605-619

mejrs replied:
> I used `PyErr::warn_bound` and that seems to do the conversion to CString and matches what you refer to as the new API :)

Right, but what I mean is that we can move this from runtime to compile time. See https://github.com/PyO3/pyo3/pull/4255 for example.

We could have a function like this:
```rust
   fn warn_with_cstr<'py>(
        py: Python<'py>,
        category: &Bound<'py, PyAny>,
        message: &CStr,
        stacklevel: i32,
    ) -> PyResult<()> {
        error_on_minusone(py, unsafe {
            ffi::PyErr_WarnEx(
                category.as_ptr(),
                message.as_ptr(),
                stacklevel as ffi::Py_ssize_t,
            )
        })
    }
```

FlickerSoul replied:
Thanks for the clarification! Moving that to compile time makes a lot sense :)

Icxolu replied:
I don't think we can parse it as `LitCStr` because of MSRV. But we can make use of the `utils::LitCStr` compatibility struct that we introduced lately to to conditionally emit either a c-string literal or `ffi::c_str!`.

FlickerSoul replied:
That makes sense. I just saw the MSRV check. Will work on that later. :)

mejrs commented on tests/ui/invalid_deprecated_pyfunction.rs:
```diff
@@ -0,0 +1,27 @@
+#[pyo3(deprecated(message = "deprecated function", random_key))]
+fn random_key_deprecated() {}
+
+fn main() {}
```
If you're developing on Windows, you need to make these files end with a blank line ;)

FlickerSoul replied:
Oh that's interesting and good to know :)

mejrs reviewed (commented):
Thanks, this looks good. A couple of thoughts:

- Reading this, it feels like this is really just a general mechanism for raising warnings, not necessarily deprecation warnings. Should we maybe just call it pyo3(warn)? 🤔 
- Should the message be optional? It seems like https://docs.python.org/3/library/warnings.html#warnings.warn allows this.

mejrs commented on guide/src/class/protocols.md:
```diff
@@ -424,6 +424,8 @@ cleared, as every cycle must contain at least one mutable reference.
   - `__traverse__(<self>, pyo3::class::gc::PyVisit<'_>) -> Result<(), pyo3::class::gc::PyTraverseError>`
   - `__clear__(<self>) -> ()`
 
+> Note: `__traverse__` does not work with [`#[pyo3(deprecated(...))]`](../function.md#deprecated) attribute.
+
```
Does this (or using it on classattr) raise a compile error? Or does it just do nothing?

FlickerSoul replied:
Ah it currently does nothing. I should have made a compile error for that!

FlickerSoul commented:
> * Reading this, it feels like this is really just a general mechanism for raising warnings, not necessarily deprecation warnings. Should we maybe just call it pyo3(warn)? 🤔

The name `pyo3(warn)` seems to make more sense! I also suggested this in the original issue but I probably didn't make that clear :)

> * Should the message be optional? It seems like https://docs.python.org/3/library/warnings.html#warnings.warn allows this.

The signature of `warn` is `warnings.warn(message, category=None, stacklevel=1, source=None, *, skip_file_prefixes=None)` so I think `message` is required. 

If we are making it `pyo3(warn)`, should we also parse for `stacklevel` in the attribute, because `PyErr::warn_bound` (internally `PyErr_WarnEx`) accepts it optionally? 

I'll work on additional testing and the compile error, and wait for more opinions on what this attribute should be called :)

Icxolu commented:
I like the idea of `#[pyo3(warn)`, but in that case I think we should match Python and default to `UserWarning`. Maybe we want to add a shortcut like `#[pyo3(deprecated = "msg")]` that would match `#[pyo3(warn(message = "msg", category = "DeprecationWarning"))]`?

davidhewitt commented:
Reading this, I also think the general mechanism `#[pyo3(warn)]` seems like the right choice. 

Probably the `category` should be a path to a Rust type, i.e. `#[pyo3(warn(message = "msg", category = PyDeprecationWarning))]`

Having been away from the discussion for a while, did we decide strongly that we don't want to support the Rust `#[deprecated]` message? My intuition is that it's still extremely convenient and intuitive to have `#[deprecated(note = "foo")]` to automatically set `#[pyo3(warn(message = "foo", category = PyDeprecationWarning)]`.

In the (I think rare) case where users want a Rust deprecation but not a Python one, we could always have `#[pyo3(warn(false))]`.

Finally, a style question. I think that `#[pyo3(warn(message = "..."))]` is the first time that we've had parenthesised pyo3 options. Should we consider e.g. `#[pyo3(warn = (message = "..."))]` or another syntax instead? I think `warn(...)` is _probably_ the right choice as we see it elsewhere in the Rust ecosystem, but I'd at least like to make sure we all are happy before we add a new style of attribute to the library.

davidhewitt commented:
> Maybe we want to add a shortcut like `#[pyo3(deprecated = "msg")]` that would match `#[pyo3(warn(message = "msg", category = "DeprecationWarning"))]`?

Personally I'd want this shorthand to just be the Rust `#[deprecated(note = "msg")]` as per my comment above.

FlickerSoul commented:
That sounds good! I will refactor the current one to `warn` and make a new `deprecated` to be a shortcut of `DeprecationWarning` :)

Icxolu commented:
> Having been away from the discussion for a while, did we decide strongly that we don't want to support the Rust `#[deprecated]` message? My intuition is that it's still extremely convenient and intuitive to have `#[deprecated(note = "foo")]` to automatically set `#[pyo3(warn(message = "foo", category = PyDeprecationWarning)]`.

There was some discussion about that in #4316. I kind of agree with the points made there, that reusing the `#[deprecated]` attribute is not such a good idea. 
- we loose control about syntax and options
- deprecating only Rust / only Python get harder
- deprecating both, but using a different category for Python gets harder 
- using different messages for Rust and Python would not be possible
-  it starts to blur the line of what only effects Rust, and what does have an effect on Python as well. `#[must_use]` would be an example that does not have any effect on Python, but could still occur on a `#[pyfunction]` or `#[pymethod]`.
-  It would also be the only option that's not configured through `#[pyo3(...)]`

I think the most straight forward solution without additional complexity or "magic" is to stick to our options system, possibly with an option, opt-in or opt-out, to also emit the `#[deprecated]`. attribute.

> Finally, a style question. I think that `#[pyo3(warn(message = "..."))]` is the first time that we've had parenthesised pyo3 options. Should we consider e.g. `#[pyo3(warn = (message = "..."))]` or another syntax instead? I think `warn(...)` is _probably_ the right choice as we see it elsewhere in the Rust ecosystem, but I'd at least like to make sure we all are happy before we add a new style of attribute to the library.

I think this (`#[pyo3(warn(message = "..."))]`) is fairly common syntax. The `#[pyo3]` attribute itself uses that syntax, so having it for nested options as well, makes sense to me.

davidhewitt commented:
That makes sense. I still think that most commonly users will want to deprecate both sides at the same time and so tying both together with Rust's `#[deprecated]`. I am happy to cede my opinion as the minority here.

Let's go with `#[pyo3(deprecated = "...")]` shorthand in that case, and keep it Python only.

FlickerSoul commented:
> I still think that most commonly users will want to deprecate both sides at the same time and so tying both together with Rust's `#[deprecated]` is convenient for the average case. 

> I think the most straight forward solution without additional complexity or "magic" is to stick to our options system, possibly with an option, opt-in or opt-out, to also emit the `#[deprecated]`. attribute.

I agree with both. I proposed in the issue that we could use something like `#[pyo3(deprecated = "...", python_only)]`, if we want to make `#[pyo3(deprecated = "...")]` deprecate both sides. Though, I can't think of a good name other than `python_only`. 

In this way, users don't need to duplicate their deprecation note twice if they want to deprecate both sides. They also have the control of which side of deprecation they'd like to apply on. This seems to be a win-win for both sides.

FlickerSoul commented:
Hi all! 

The `#[pyo3(warn(message = "...", category = ...))]` and `#[pyo3(deprecated = "...")]` should have been implemented properly. 

The `warn` attribute now behaves like the `warnings.warn` and emits `UserWarning` when category is absent. 

The `deprecated` attribute currently only deprecates Python side. If we want to allow `deprecated` to deprecate both sides and add an opt-out option, like the `#[pyo3(deprecated = "...", python_only)]`, I can add it later this week. 

Please let me know if I missed anything. Thanks! 

Best regards,
L

Icxolu reviewed (commented):
Thanks! This mostly looks good to me. Just some minor things here and there.

I saw that you opened this to support multiple warnings. I currently don't see a big problem with that. It _does_ however interact with automatic Rust deprecation since we currently could have multiple `#[pyo3(deprecated)]`, but only one `#[deprecated]` attribute is allowed per item, so we should be aware of the consequences before committing to anything here. Let's see what the others think of this. This would currently be the only pyo3 option that would be allowed multiple times (I think), so we should document that behavior somewhere if we decide to keep it.

I think we should add one or two of these attributes to the `tests/hygiene` tests, to make sure all the emitted code is fully qualified.

Also it should be possible to use a user defined warning category that extends from `PyWarning`. Maybe we should add a test to make sure that works?

> The deprecated attribute currently only deprecates Python side. If we want to allow deprecated to deprecate both sides and add an opt-out option, like the #[pyo3(deprecated = "...", python_only)], I can add it later this week.

I'm open to this, but I would suggest if we decide to go down that route we should do it as a followup.

FlickerSoul commented on pyo3-macros-backend/src/pyfunction.rs:
```diff
@@ -83,13 +85,164 @@ impl PyFunctionArgPyO3Attributes {
+    pub span: Span,
+}
+
+pub enum PyFunctionWarningCategory {
+    Path(Path),
+    DeprecationWarning,
+}
```
That's a great idea!

FlickerSoul commented on tests/ui/invalid_pymethods_deprecated.rs:
Ahhh yes!! Thanks for catching that!

FlickerSoul commented:
> I saw that you opened this to support multiple warnings. I currently don't see a big problem with that. It _does_ however interact with automatic Rust deprecation since we currently could have multiple `#[pyo3(deprecated)]`, but only one `#[deprecated]` attribute is allowed per item, so we should be aware of the consequences before committing to anything here. Let's see what the others think of this.

I agree we should be cautious on the multiplicity of the same attribute. I'm also curious what others think of this. :)

FlickerSoul commented:
Clippy fails on x86_64 linux target surprisingly, because of 

```
  error: struct `PyFunction` is never constructed
     --> src/types/function.rs:184:12
      |
  184 | pub struct PyFunction(PyAny);
      |            ^^^^^^^^^^
      |
      = note: `-D dead-code` implied by `-D warnings`
      = help: to override `-D warnings` add `#[allow(dead_code)]`
```

It seems that `PyFunction` is used with `ffi` and is not intended to be constructed manually. Should I mark it with `#[allow(dead_code)]`?

FlickerSoul commented:
> Only on beta though. I wonder if that's a newly introduced rustc bug?


Looks like so?

The check had been passing until this one. Could you maybe rerun the check to see if the error goes away?

Icxolu commented:
This is just dead code detection getting better 🎉. It also happens on main. The reason is our reexport of `PyFunction` 
https://github.com/PyO3/pyo3/blob/b2b5203ca4fb103e84f1ae8a267ea2d2912d5307/src/types/mod.rs#L25-L26
has a more restrictive gate than its type definition
https://github.com/PyO3/pyo3/blob/b2b5203ca4fb103e84f1ae8a267ea2d2912d5307/src/types/function.rs#L183-L184
So that type is actually not nameable on PyPy 3.8+ but it was actually intended to be.

You don't have to worry about it here, we can fix that separately. I'll file a PR shortly.

FlickerSoul commented:
Awesome!! Thanks!

FlickerSoul commented:
@miikkas That's great suggestion! If I understand the PEP 702, `__deprecated__` is mainly type checkers. I would open a separate PR for that as the changes in this PR are getting bigger. 

Would it be helpful if I squash the commits now, because this thread is getting long? 

@Icxolu All requested changes have been made and thanks for the detailed feedback! I also rebased on main to resolve the clippy error. It looks like nightly compiler is also getting better and emits new errors. I can open a separate PR to fix it. :)

I also see that coverage is not hitting the target. They are mainly caused by functions in the `ToTokens` trait of the attribute struct and a couple helper functions to get `Span`, required by others parts of code. The functions are only called when errors are raised and it looks like `test_compile_error` won't help with the coverage. Could you suggest a place to add tests for those?

Icxolu commented:
> All requested changes have been made and thanks for the detailed feedback! I also rebased on main to resolve the clippy error.

@FlickerSoul Thanks, implementation and tests look good to me. I don't think we need to worry too much about CodeCov not hitting these specific paths.

> It looks like nightly compiler is also getting better and emits new errors. I can open a separate PR to fix it. :)

Please feel free to do so 😄 

> Would it be helpful if I squash the commits now, because this thread is getting long?

I leave that up to you, we have squash merging enabled anyway, so from that perspective it does not really matter. It stretches the thread a bit, but depending on your workflow the commits could be useful if we device to take a step back, more on that below.

> Thanks, this looks great! One question: Would adding the `__deprecated__` attribute to the Python object (as per [PEP 702](https://peps.python.org/pep-0702/#runtime-behavior)) be out of scope?

@miikkas Thanks for bringing this to our attention. I can definitely see us supporting that. However I agree with @FlickerSoul that we should postpone that to a followup. This means we should carefully evaluate what we can and want to guarantee in this first base implementation.

From my reading PEP 702 talks specifically about deprecations, and not warnings in general. This could force us to diverge `#[pyo3(warn)]` and `#[pyo3(deprecated)]` in the future, since matching on item names (in this case the `category`) from macro code is usually not a good idea. This changes a bit what we want to guarantee here about the equivalence between them. 

This together with a possibility of automatic Rust `#[deprecated]` makes me kind of lean towards only supporting a single warning (or at least only a single `#[pyo3(deprecated)]` xor multiple `#[pyo3(warn)]`) for now. (Multiple `#[pyo3(deprecated)]` would not make a lot of sense anyway IMHO) We can always open it up later once we put more design work into either or both of these directions.

cc @davidhewitt A second opinion here would be appreciated 🙃

FlickerSoul commented on guide/src/function.md:
```diff
@@ -96,6 +98,120 @@ The `#[pyo3]` attribute can be used to modify properties of the generated Python
+  - <a id="deprecated"></a> `#[pyo3(deprecated = "...")]`
+
+    Set this option to display deprecation warning when the function is called in Python. 
+    This is equivalent to [`#[pyo3(warn(message = "...", category = PyDeprecationWarning))]`](#warn) or [`warnings.warn(message, DeprecationWarning)`](https://docs.python.org/3.12/library/warnings.html#warnings.warn).
```
Yeah I agree this should be changed. I wrote this before reading the PEP 702. The semantics of `deprecated` in the PEP doesn't seem to match what I'm promising here. Using the same name thus could cause future confusion.

FlickerSoul commented on guide/src/function.md:
````diff
@@ -96,6 +98,120 @@ The `#[pyo3]` attribute can be used to modify properties of the generated Python
+
+    ```python
+    import warnings
+    from module_with_deprecated_fn import function_with_warning, function_with_warning_and_custom_category
````
I should be more careful with the changes 🥲. Thanks for catching that!

FlickerSoul commented:
> This together with a possibility of automatic Rust `#[deprecated]` makes me kind of lean towards only supporting a single warning (or at least only a single `#[pyo3(deprecated)]` xor multiple `#[pyo3(warn)]`) for now. (Multiple `#[pyo3(deprecated)]` would not make a lot of sense anyway IMHO) We can always open it up later once we put more design work into either or both of these directions.

I agree multiple `#[pyo3(deprecated)]` wouldn't make sense and I like the idea of allowing the ability to use multiple `#[pyo3(warn)]`. Would "at least only a single `#[pyo3(deprecated)]` (logical) or multiple `#[pyo3(warn)]`" be a viable option? It seems to me this way we give users more choices without going to trouble. 

I'll wait on @davidhewitt's opinion before going for the implementation. :))


> This could force us to diverge `#[pyo3(warn)]` and `#[pyo3(deprecated)]` in the future, since matching on item names (in this case the `category`) from macro code is usually not a good idea. This changes a bit what we want to guarantee here about the equivalence between them.

This makes sense. I can't think of a good API for this off the top of my head. Maybe David can shed some light on this as well :)) 

Again, thanks for the feedback. Much appreciated!

mejrs commented:
My apologies for the delay. To me personally the PR looks good and I'd like to merge it. @Icxolu was your review addressed?

Finally, can you please resolve the merge conflicts?

Icxolu commented:
Thanks for bringing this up again @mejrs. I think implementation wise this looked good to me as well. I think the last discussion point was the guarantee about the equivalency of `#[pyo3(deprecated(...))]` and `#[pyo3(warn(message = "...", category = PyDeprecationWarning))]``. If we want to support PEP 702 in the future, we should remove this guarantee from the docs, and maybe even emphasize that they may _not_ be equivalent. Alternatively we could also split this, land the general warning mechanism first, and add specialized deprecations later.

mejrs commented:
> Alternatively we could also split this, land the general warning mechanism first, and add specialized deprecations later.

That sounds good to me. I propose we remove `#[pyo3(deprecated)]` from this pr, and just do `#[pyo3(warn)]` for now.

FlickerSoul commented:
Sounds good. I'll try to get back to you by this week. :)

FlickerSoul commented:
Hi @mejrs and @Icxolu,

I removed the `#[pyo3(deprecated)]` from this PR and updated the code and doc to only have `#[pyo3(warn)]`. 

I noticed that the current `#[pyo3(deprecated)]` only supports the first two arguments of the [`warn`](https://docs.python.org/3.13/library/warnings.html#warnings.warn) function in `warnings`, namely `message` and `category`. Do you think if we want to add support for the rest three arguments: `stacklevel`, `source` and `skip_file_prefixes`? 

Additionally, the UI tests are failing for Ubuntu but pass on my Mac. What should I do about those?

mejrs commented:
Can you edit the original post of this PR too? To avoid confusing future readers?

> I noticed that the current `#[pyo3(deprecated)]` only supports the first two arguments of the [`warn`](https://docs.python.org/3.13/library/warnings.html#warnings.warn) function in `warnings`, namely `message` and `category`. Do you think if we want to add support for the rest three arguments: `stacklevel`, `source` and `skip_file_prefixes`?

I'd err on the side of not having this. It keeps the documentation simple and we can always add it later. I don't think any of them are particularly useful (from native code, anyway) and users can always just raise warnings manually if they really need it.

> Additionally, the UI tests are failing for Ubuntu but pass on my Mac. What should I do about those?

It's just error output from the compiler that has changed, you should update to the latest stable compiler. And bless the output, with the environment variable TRYBUILD=overwrite

Icxolu commented on pyo3-macros-backend/src/pyfunction.rs:
```diff
@@ -83,13 +86,157 @@ impl PyFunctionArgPyO3Attributes {
+pub enum PyFunctionWarningCategory {
+    Path(Path),
+    UserWarning,
+    DeprecationWarning,
```
I assume this one is unused for now. It's probably no big deal to leave it in, but a `todo` or `fixme` comment with the intend would be nice.

Icxolu reviewed (approved):
Once CI is green this looks good to me 🚀 Thanks @FlickerSoul for the work and all the patience.

(There is also `nox -s update-ui-tests` to bless the tests)

davidhewitt commented:
For sake of merging this before #5107, I'll push a fix for the ui tests now.

davidhewitt commented:
... actually I'm not able to do so (looks like permissions have not been given for maintainers to push to the branch).

FlickerSoul commented:
Hi all, sorry for the delay. I'll push a fix to the UI tests and resolve the conflicts today :))

FlickerSoul commented:
Hi @davidhewitt, I'm unable to fix the UI test even though I'm using the latest stable rust compiler. I have enabled allow editing by maintainers. Could you commit the patch for the UI tests? Thanks!

mejrs commented on tests/ui/invalid_intopy_with.stderr:
```diff
@@ -7,8 +7,8 @@ error[E0308]: mismatched types
 5 |     #[pyo3(into_py_with = into)]
```
You're probably using an older compiler than CI. In any case, I think you can drop this changed file from the PR and it should just all work.

Icxolu commented:
I've excluded the tests from the free-threaded build and added some comments. I think this is ready to land now, thanks to everyone involved!
